# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [2]:
# importing pandas and the dataset

import pandas as pd
insurance_df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")

In [7]:
# Identify the dimensions of the dataset by determining the number of rows and columns it contains.

insurance_df.shape
print("The dataset contains",insurance_df.shape[0], "rows and",insurance_df.shape[1], "columns")

The dataset contains 4008 rows and 11 columns


In [24]:
# Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. 

insurance_df.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

**Evaluate whether the data type of each column is appropriate:**

Customer                      object -> dtype is appropriate, I expect unique strings<br>
ST                            object -> dtype is appropriate, I expect categorical strings<br>
GENDER                        object -> dtype is appropriate, I expect categorical strings<br>
Education                     object -> dtype is appropriate, I expect categorical strings<br>
Customer Lifetime Value       object -> dtype is a surprise, I expect numerical data<br>
Income                       float64 -> dtype is appropriate, I expect numerical floats<br>
Monthly Premium Auto         float64 -> dtype is appropriate, I expect numerical floats<br>
Number of Open Complaints     object -> dtype is a surprise, I expect numerical data (integer)<br>
Policy Type                   object -> dtype is appropriate, I expect categorical strings<br>
Vehicle Class                 object -> dtype is appropriate, I expect categorical strings<br>
Total Claim Amount           float64 -> dtype is appropriate, I expect numerical floats<br>

**Provide suggestions for fixing any incorrect data types:**
   - I would expect CLV to be a number
   - I would also expect number of complaints to be an integer

In [36]:
# Identify the number of unique values for each column and determine which columns appear to be categorical. 

print(insurance_df[insurance_df.columns].nunique())

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64


In [49]:
# Cat var: Customer, ST, GENDER, Education, Policy Type, Vehicle Class
# not important cat var: Customer & GENDER 
# relevant_cat_var = ["ST","Education","Policy Type","Vehicle Class"]

# Num var: Income, Monthly Premium Auto, Total Claim Amount
# Customer Lifetime Value & Number of Open Complaints are strings but should be numerical values

relevant_cat_var = ["ST","Education","Policy Type","Vehicle Class"]
for col in relevant_cat_var:
    print(insurance_df[col].value_counts())
    print(" ")

ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64
 
Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64
 
Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64
 
Vehicle Class
Four-Door Car    576
Two-Door Car     205
SUV              199
Sports Car        57
Luxury SUV        20
Luxury Car        14
Name: count, dtype: int64
 


**Desscribe the unique values of each categorical column and the range of values for numerical columns, and give your insights.**

Column name: ST  
Value range: 8  
Description: Only a few US states were listed  
Insights: This insurance company seems to be based in the Western region of US  

Column name: Education  
Value range: 6  
Description: A categorical list of different education levels of the policy holders  
Insights: The majority of policy holders have an educational level of Bachelor's or below  

Column name: Policy Type  
Value range: 3   
Description: A categorical list of the different available policy types  
Insights: The majority of policy holders have a policy for their personal car  

Column name: Vehicle Class  
Value range: 6  
Description: A categorical list of car-type options  
Insights: The majority of policy holders have a 'regular' two- or four-door car  
    
    

In [43]:
insurance_df.head()

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

**Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the numerical data**   

In [45]:
insurance_df["Customer Lifetime Value"].values

array([nan, '697953.59%', '1288743.17%', ..., nan, nan, nan], dtype=object)

In [46]:
insurance_df["Number of Open Complaints"].values

array(['1/0/00', '1/0/00', '1/0/00', ..., nan, nan, nan], dtype=object)

In [48]:
insurance_df.describe()

Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678

**Provide your conclusions based on these summary statistics**   
 - The data have too many decimal places and needs to be rounded up
 - Two numerical series are rendered unintelligible due to having strings as values (CLV & Number of Open Complaints)
 - The mean income of poliy holders is higher than the mean monthly premium
 - The mean total claim amount is lower than the average monthly premium collected (1071 customers x mean monthly premium): meaning more comes in than is paid out (see max of monthly premium vs total claim amount)

**Compute summary statistics for categorical columns and providing your conclusions based on these statistics**
 - The dataset appears to only hold a third of values (1071 non-null values in a dataset of 4008 rows)
 - Mode for Education is Bachelor with a frequency of 324 out of 1071
 - Most popular state among policy holders is Oregon
 - Most insured car is a four-door car
 - Most commom policy is a personal auto policy

In [52]:
#relevant_cat_var = ["ST","Education","Policy Type","Vehicle Class"]
for col in relevant_cat_var:
    print(insurance_df[col].describe())
    print(" ")

count       1071
unique         8
top       Oregon
freq         320
Name: ST, dtype: object
 
count         1071
unique           6
top       Bachelor
freq           324
Name: Education, dtype: object
 
count              1071
unique                3
top       Personal Auto
freq                780
Name: Policy Type, dtype: object
 
count              1071
unique                6
top       Four-Door Car
freq                576
Name: Vehicle Class, dtype: object
 


In [53]:
# Personal notes | variables
# Cat var: Customer, ST, GENDER, Education, Policy Type, Vehicle Class  
# Num var: Income, Monthly Premium Auto, Total Claim Amount

# Customer Lifetime Value       object   
# Number of Open Complaints     object   

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [59]:
insurance_df["ST"].value_counts().sort_values().head()

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [62]:
insurance_df["Policy Type"].value_counts().sort_values(ascending=False).idxmax()

'Personal Auto'

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [65]:
insurance_df.loc[insurance_df["Policy Type"]=="Personal Auto","Income"].mean()

38180.69871794872

In [66]:
insurance_df.loc[insurance_df["Policy Type"]=="Corporate Auto","Income"].mean()

41390.31196581197

The average Corporate Auto policy holder has a higher avg. income than the Personal Auto policy holders

### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

**Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile**

In [69]:
high_policy_claimants = insurance_df.loc[insurance_df["Total Claim Amount"]>532.8]

**Use DataFrame methods to calculate summary statistics about the high policy claim amount data.**

In [71]:
high_policy_claimants.describe()

Income  Monthly Premium Auto  Total Claim Amount
count    264.000000            264.000000          264.000000
mean   23677.344697            165.193182          782.228263
std    27013.483721            623.930992          292.751640
min        0.000000             63.000000          537.600000
25%        0.000000             99.000000          606.521741
50%    18807.000000            114.000000          679.597985
75%    42423.750000            133.250000          851.400000
max    99316.000000          10202.000000         2893.239678